src/basic/terminal-util.c

```diff
@@ -1206,30 +1208,52 @@ bool terminal_is_dumb(void) {
-         * we are PID 1 then we do not check whether we are connected to a TTY, because we don't keep /dev/console open
-         * continuously due to fear of SAK, and hence things are a bit weird. */
-
-        if (cached_colors_enabled < 0) {
-                int val;
-
-                val = getenv_bool("SYSTEMD_COLORS");
-                if (val >= 0)
-                        cached_colors_enabled = val;
-
+color_mode get_color_mode(void) {
+
+        /* Returns the mode used to choose output colors. The possible modes are COLOR_OFF for no colors,
+         * COLOR_16 for only the base 16 ANSI colors, COLOR_256 for more colors and COLOR_ON for unrestricted
+         * color output. For that we check $SYSTEMD_COLORS first (which is the explicit way to
+         * change the mode). If that didn't work we turn colors off unless we are on a TTY. And if we are on a TTY
+         * we turn it off if $TERM is set to "dumb". There's one special tweak though: if we are PID 1 then we do not
+         * check whether we are connected to a TTY, because we don't keep /dev/console open continuously due to fear
+         * of SAK, and hence things are a bit weird. */
+
+        if (cached_color_mode < 0) {
+                char *e;
+
+                e = getenv("SYSTEMD_COLORS");
+                if (e) {
+                        if (streq(e, "16"))
+                                cached_color_mode = COLOR_16;
+                        else if (streq(e, "256"))
+                                cached_color_mode = COLOR_256;
+                        else if (parse_boolean(e))
+                                cached_color_mode = COLOR_ON;
+                        else if (!parse_boolean(e))
+                                cached_color_mode = COLOR_OFF;
+                        else
+                                cached_color_mode = _COLOR_INVALID;
+                }
                 else if (getenv("NO_COLOR"))
                         /* We only check for the presence of the variable; value is ignored. */
-                        cached_colors_enabled = false;
+                        cached_color_mode = COLOR_OFF;
 
                 else if (getpid_cached() == 1)
                         /* PID1 outputs to the console without holding it open all the time */
-                        cached_colors_enabled = !getenv_terminal_is_dumb();
+                        cached_color_mode = getenv_terminal_is_dumb() ? COLOR_OFF : COLOR_16;
                 else
-                        cached_colors_enabled = !terminal_is_dumb();
+                        cached_color_mode = terminal_is_dumb() ? COLOR_OFF : COLOR_256;
+        }
+
+        return cached_color_mode;
+}
+
+bool colors_enabled(void) {
+
+        /* Returns true if colors are considered supported on our stdout. */
+
+        if (cached_colors_enabled < 0) {
+              cached_colors_enabled = get_color_mode() > 0;
         }
```

Comment:
Please drop unnecessary `{}`. Also, please fix the indentation. We use 8ch indentation.

BTW, I think it is not necessary to cache `colors_enabled`. How about to introduce an inline function in terminal-util.h:
```
static inline bool colors_enabled(void) {
        return get_color_mode() > 0;
}
```

Reply:
> Please drop unnecessary {}. Also, please fix the indentation. We use 8ch indentation.

Sorry, I messed this up. Fixed.

> BTW, I think it is not necessary to cache colors_enabled

Wouldn't this result in `get_color_mode` being executed much more frequently?
Also, what about `underline_enabled`? That it also cached.